### ISSUE TITLE
[bug] Concurrency - installing same package from multiple workers seems to generate races

### ISSUE THREAD

**Issue by fulara** [conan-io/conan#6753]
### Environment:
Preconditions: Conan is already installed and is happily sitting in some directory. 
Conan is packaged using pyinstaller. 
* Red Hat Enterprise Linux Server release 6.10 (Santiago)
* python 3.6.10
* conan 1.23.0
* compiler: gcc 8.2

### Steps 
What are we doing.
We are launching multiple workers that install 5 instances of the same app, so 5 workers invoke something like:
`conan install reference -if .... -g ..`
Immediatelly after `conan install` finished we are  sourcing envs and launching the apps. 

What we have observed is the following:
There are instances of the app that fail to boot with error like this:
```
app: error while loading shared libraries: libboost_wave.so.1.72.0: cannot open shared object file: No such file or directory
```
So first instance that installed and tried to start failed, however the further instances were fine.
This looks like the first worker have not properly recognized whether that  package `boost`is still being downloaded, and have not waited for the download to finish ( maybe they thought the download has succeded already )?

In the logs I can see some like these:
```
openssl/1.0.2t is locked by another concurrent conan process, wait...
If not the case, quit, and do 'conan remove --locks'
```
But there is no such entry for `boost`. (which of course can happen, but still.. )

We have seen this issue on 2 occassions out of 7 clear installs we have done so far. 
Any thoughts?
What sort of logs would you require from us?

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by jgsogo** [comment]
Hi, @fulara!

We are trying to implement concurrency at some points, but is something that we are not supporting right now... nevertheless, `conan install` should work in parallel with that kind of locks your show in the output.

I think that logs won't help, because we'd need to know the order they are printed into different terminals (and it can be very confusing). The way to go is to try to reproduce the scenario consistently and look to the sources to see where those locks are working and which operations that might interfere are not protected by a lock.

So, anything regarding your environment will be useful, or any insight/guess you might have will help: 
 * all jobs download the same package ID?
 * are you using the download cache?
 * can you check the generated files (which generator you are using)? I'm looking for missing directories, there is a `filter_empty` parameter inside the code that can be problematic.

This is going to be hard, but it will help to improve the concurrency 💪

**Comment by fulara** [comment]
Hello @jgsogo !
I've chatted with @memsharded  about this - and he says that this particular area is something that he is thinking of improving. in 1.25 

Regarding:
> We are trying to implement concurrency at some points, but is something that we are not supporting right now...

your [documentation](https://docs.conan.io/en/latest/mastering/custom_cache.html#concurrency) states:
```
Conan local cache support some degree of concurrency, allowing simultaneous creation or installation of different packages
```

back to your questions:
* all jobs download the same package ID?  
Yes, they all download same package.
* are you using the download cache?
No, just single local cache thats empty initially, 
* can you check the generated files (which generator you are using)? I'm looking for missing directories, there is a filter_empty parameter inside the code that can be problematic.
Not sure what exactly you are asking for here, what sort of filter_empty? To me this issue looks like I am purely guessing here: 
one instance of  `conan` goes ahead and starts installing the boost/_/_ directory, however it has not yet created the lockfie. 
Second instance sees that boost/_/_ already exists and it does not contains the lock file so it assumes that everything is downloaded. and just moves on.
we are using -g virtualrunenv and -g virtualenv

**Comment by jgsogo** [comment]
> * can you check the generated files (which generator you are using)? I'm looking for missing directories, there is a filter_empty parameter inside the code that can be problematic.
>>   Not sure what exactly you are asking for here, what sort of filter_empty? To me this issue looks like I am purely guessing here:
>>   one instance of  `conan` goes ahead and starts installing the boost/_/_ directory, however it has not yet created the lockfie.
>>   Second instance sees that boost/_/_ already exists and it does not contains the lock file so it assumes that everything is downloaded. and just moves on.
>>   we are using -g virtualrunenv and -g virtualenv

Are the files generated by `virtualrunenv` and `virtualenv` the same for all the jobs? Can you check the content of the `environment.xxxx` ones?

The `filter_empty` is a variable we have inside the sources (read below).

My guess: 4 jobs wait in the lock while one of them is working downloading and installing `boost` package. All of them are [waiting here](https://github.com/conan-io/conan/blob/7e88e8b35a7519c3ec6d20bb7f09817c46961d6d/conans/client/installer.py#L457) **where they have already decided they need to download `boost`** (maybe they should check again), then every job will [remove the folder to copy the package](https://github.com/conan-io/conan/blob/7e88e8b35a7519c3ec6d20bb7f09817c46961d6d/conans/client/remote_manager.py#L135) and here we can have a race condition: one job has removed the folder for the package while the other one is [filtering the folders in the package](https://github.com/conan-io/conan/blob/7e88e8b35a7519c3ec6d20bb7f09817c46961d6d/conans/model/build_info.py#L68) (here you have the `filter_empty`) before printing them in the file generated by the generators. This job removes some folders, and you cannot find your shared libraries.

I might be wrong, I'm just jumping from one file to another in Github, but it is plausible for me.

Do you have a  better guess, @memsharded ?

**Comment by fulara** [comment]
I dont think they are waiting for `boost` to download, they think its already downloaded. as there is no message:
`openssl/1.0.2t is locked by another concurrent conan process, wait...` for `boost`.

So I did more more tests, and here are the results:
<details>
 <summary>run 1) crashed with stack trace - all instances started though!</summary>
that particular worker had crash inside my-package installation but it exited that and started fine.
<pre><code>
my-package/3.4 is locked by another concurrent conan process, wait...
If not the case, quit, and do 'conan remove --locks'
Traceback (most recent call last):
  File "conan/conans/client/command.py", line 1974, in run
  File "conan/conans/client/command.py", line 500, in install
  File "conan/conans/client/conan_api.py", line 81, in wrapper
  File "conan/conans/client/conan_api.py", line 521, in install_reference
  File "conan/conans/client/manager.py", line 47, in deps_install
  File "conan/conans/client/graph/graph_manager.py", line 114, in load_graph
  File "conan/conans/client/graph/graph_manager.py", line 240, in _resolve_graph
  File "conan/conans/client/graph/graph_manager.py", line 344, in _load_graph
  File "conan/conans/client/graph/graph_builder.py", line 66, in load_graph
  File "conan/conans/client/graph/graph_builder.py", line 126, in _expand_node
  File "conan/conans/client/graph/graph_builder.py", line 228, in _expand_require
  File "conan/conans/client/graph/graph_builder.py", line 126, in _expand_node
  File "conan/conans/client/graph/graph_builder.py", line 201, in _expand_require
  File "conan/conans/client/graph/graph_builder.py", line 401, in _create_new_node
  File "conan/conans/client/graph/graph_builder.py", line 369, in _resolve_recipe
  File "conan/conans/client/graph/proxy.py", line 33, in get_recipe
  File "conan/conans/client/graph/proxy.py", line 54, in _get_recipe
  File "conan/conans/paths/package_layouts/package_cache_layout.py", line 172, in load_metadata
  File "conan/conans/model/package_metadata.py", line 95, in loads
  File "json/__init__.py", line 354, in loads
  File "json/decoder.py", line 339, in decode
  File "json/decoder.py", line 357, in raw_decode
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
</code>
</details>
run 2 - ok, no issues
 <summary>run 3 - same issue again.</summary>
<details>
<pre><code>
error while loading shared libraries: libboost_wave.so.1.72.0: cannot open shared object file: No such file or directory
</code></pre>

I have compared files: activate_run.sh activate.sh deactivate_run.sh deactivate.sh environment_run.sh.env environment.sh.env
Yes, they are the same between working version and not.
</details>

Seems pretty reproducible to me, so I do more tests for you if you want.

**Comment by jgsogo** [comment]
The log you have posted is a failure reading a `metadata.json` file, some other process might be writing it at the same time. Yes, here we have a race and I'm sure it is not the only one. We need to work on Conan concurrency to fix existing issues and to add more concurrency.

**Comment by memsharded** [comment]
Closing in favor of #15840 where the Conan 2 cache concurrency future progress will be reported.
